packages/contracts/contracts/L1/ValidatorPool.sol

```diff
@@ -104,73 +111,233 @@ contract ValidatorPool is ReentrancyGuardUpgradeable, Semver {
+                address lastValidator = validators[lastValidatorIndex];
 
-        if (validatorDeposits < SLASHING_AMOUNT) {
-            slashing = validatorDeposits - SLASHING_AMOUNT;
+                validators[validatorIndex] = lastValidator;
+                validatorIndexes[lastValidator] = validatorIndex;
+            }
+            delete validatorIndexes[_validator];
+            validators.pop();
         }
 
-        depositsOf[validator] -= slashing;
-        depositsOf[challenger] += slashing;
+        balances[_validator] = balance;
+    }
 
-        emit Slashing(validator, challenger, validatorDeposits);
+    /**
+     * @notice Returns the bond corresponding to the output index. Reverts if the bond does not exist.
+     *
+     * @param _outputIndex Index of the L2 checkpoint output.
+     *
+     * @return The bond data.
+     */
+    function getBond(uint256 _outputIndex) external view returns (Types.Bond memory) {
+        Types.Bond storage bond = bonds[_outputIndex];
+        require(bond.amount > 0 && bond.expiresAt > 0, "ValidatorPool: the bond does not exist");
+        return bond;
     }
 
     /**
-     * @notice Returns the deposit balance of validator.
+     * @notice Returns the balance of given address.
      *
-     * @param _validator The address of validator
+     * @param _addr Address of validator.
      *
-     * @return Deposit balance of validator.
+     * @return Balance of given address.
      */
-    function balanceOf(address _validator) public view returns (uint256) {
-        return depositsOf[_validator];
+    function balanceOf(address _addr) external view returns (uint256) {
+        return balances[_addr];
     }
 
-    function isValidator(address _validator) public view returns (bool) {
-        return validators[_validator] || _validator == TRUSTED_VALIDATOR;
+    /**
+     * @notice Determines whether the given address is an active validator.
+     *
+     * @param _addr Address of validator.
+     *
+     * @return Whether the given address is an active validator.
+     */
+    function isValidator(address _addr) public view returns (bool) {
+        if (validators.length == 0) {
+            return false;
+        } else if (_addr == address(0)) {
+            return false;
+        }
+
+        uint256 index = validatorIndexes[_addr];
+        return validators[index] == _addr;
     }
 
-    function isValidatorAlive(address _validator) public view returns (bool) {
-        return depositsOf[_validator] > 0 || _validator == TRUSTED_VALIDATOR;
+    /**
+     * @notice Returns the number of validators.
+     *
+     * @return The number of validators.
+     */
+    function validatorCount() external view returns (uint256) {
+        return validators.length;
+    }
+
+    /**
+     * @notice Determines who can submit the L2 output next.
+     *
+     * @return The address of the validator.
+     */
+    function nextValidator() public view returns (address) {
```

Comment:
When there is no validator, we probably should return `TRUSTED_VALIDATOR`.

Reply:
Validators can't not exist because `TRUSTED_VALIDATOR` must also make deposits. If the `TRUSTED_VALIDATOR` has no deposits, no one can submit an output.

Reply:
OK good. Then, it's important to keep the `TRUSTED_VALIDATOR` eligible for validator.